### PR TITLE
[Clang][ASTMatcher] Add `dependentNameType` Matcher

### LINKED ISSUE #121240: Missing AST matcher: dependentNameType
Another matcher I realized is missing is one for types of kind `DependentNameType`.

### PR DESCRIPTION
**Pull request by AmrDeveloper**
Add AST Matcher for `DependentNameType`

Fixes: https://github.com/llvm/llvm-project/issues/121240

### CODE REVIEW

File: clang/lib/ASTMatchers/Dynamic/Registry.cpp [hunk 1]
```diff
@@ -218,6 +218,7 @@ RegistryMaps::RegistryMaps() {
   REGISTER_MATCHER(cxxTryStmt);
   REGISTER_MATCHER(cxxUnresolvedConstructExpr);
   REGISTER_MATCHER(decayedType);
+  REGISTER_MATCHER(dependentNameType);
```
[the remark this answers is not in the document]
Comment (AmrDeveloper) [reply clang/lib/ASTMatchers/Dynamic/Registry.cpp]: Done, thank you

File: clang/include/clang/ASTMatchers/ASTMatchers.h [hunk 2]
```diff
@@ -7711,6 +7711,16 @@ AST_MATCHER_P(DecayedType, hasDecayedType, internal::Matcher<QualType>,
   return InnerType.matches(Node.getDecayedType(), Finder, Builder);
 }
 
+/// Matches dependent name type
```
[the remark this answers is not in the document]
Comment (AmrDeveloper) [reply clang/include/clang/ASTMatchers/ASTMatchers.h]: Done

File: clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp [hunk 3]
```diff
@@ -1912,6 +1912,21 @@ TEST_P(ASTMatchersTest, DeducedTemplateSpecializationType) {
               deducedTemplateSpecializationType()));
 }
 
+TEST_P(ASTMatchersTest, DependentNameType) {
+  if (!GetParam().isCXX()) {
+    // FIXME: Add a test for `dependentNameType()` that does not depend on C++.
```
Comment (HighCommander4) [inline clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp:1917]: Similar to `DependentScopeDeclRefExpr`, `DependentNameType` only exists in C++, so this FIXME comment can be removed.

File: clang/docs/LibASTMatchersReference.html [hunk 4]
```diff
@@ -2536,6 +2536,15 @@ <h2 id="decl-matchers">Node Matchers</h2>
   matches "decltype(i + j)"
 </pre></td></tr>
 
+<tr><td>Matcher&lt;<a href="https://clang.llvm.org/doxygen/classclang_1_1Stmt.html">Stmt</a>&gt;</td><td class="name" onclick="toggle('dependentNameType0')"><a name="dependentNameType0Anchor">dependentNameType</a></td><td>Matcher&lt;<a href="https://clang.llvm.org/doxygen/classclang_1_1DependentNameType.html">DependentNameType</a>&gt;...</td></tr>
```
Comment (HighCommander4) [inline clang/docs/LibASTMatchersReference.html:2539]: The first column here should be `Matcher<Type>`, not `Matcher<Stmt>`

File: clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp (hunk 3 above)
Comment (AmrDeveloper) [reply clang/unittests/ASTMatchers/ASTMatchersNodeTest.cpp]: Done

File: clang/docs/LibASTMatchersReference.html (hunk 4 above)
Comment (AmrDeveloper) [reply clang/docs/LibASTMatchersReference.html]: Thank you, Done

**Review by HighCommander4 — APPROVED** [review]
LGTM, thanks!

**Comment by HighCommander4** [comment]
@AmrDeveloper if you're not tired of these yet, I have one final one: https://github.com/llvm/llvm-project/issues/121307

**Comment by AmrDeveloper** [comment]
> ~@AmrDeveloper if you're not tired of these yet, I have one final one: #121307~
> 
> (Update: that one's been taken by another contributor.)

Thank you, if you found another one, I will be interested to work on it, you can assign me directly :D

**Pull request merged**